lib/daru/dataframe.rb

```diff
@@ -906,19 +906,18 @@ def filter_vector vec
     def filter_rows &block
       return to_enum(:filter_rows) unless block_given?
 
-      df = Daru::DataFrame.new({}, order: @vectors.to_a)
-      marked = []
-
-      @index.each do |index|
-        keep_row = yield access_row(index)
-        marked << index if keep_row
-      end
-
-      marked.each do |idx|
-        df.row[idx] = self[idx, :row]
+      values = to_a[0].map{|h| h.values}
+      index = @index.to_a
+
+      index.reverse.each do |i|
+        keep_row = yield access_row(i)
+        unless keep_row
+          index.delete_at(i)
```

Comment:
Over here you're assuming that the index is strictly numerical (consists only of integers from 0..n-1). If it turns out to be something else like a string, this will fail.